components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+            .try_query_row(
+                &sql,
+                &[],
+                |row| -> rusqlite::Result<_> { Ok(row.get_checked::<_, bool>(0)?) },
+                false,
+            )?
+            .unwrap_or(false))
+    }
+
+    /// Builds a temporary table with the merge states of all nodes in the merged
+    /// tree, then updates the local tree to match the merged tree.
+    ///
+    /// Conceptually, we examine the merge state of each item, and either leave the
+    /// item unchanged, upload the local side, apply the remote side, or apply and
+    /// then reupload the remote side with a new structure.
+    fn update_local_items<'t>(
+        &self,
+        descendants: Vec<MergedDescendant<'t>>,
+        deletions: Vec<Deletion>,
+    ) -> Result<()> {
+        // First, insert rows for all merged descendants.
+        sql_support::each_sized_chunk(
+            &descendants,
+            sql_support::default_max_variable_number() / 4,
+            |chunk, _| -> Result<()> {
+                // We can't avoid allocating here, since we're binding four
+                // parameters per descendant. Rust's `SliceConcatExt::concat`
+                // is semantically equivalent, but requires a second allocation,
+                // which we _can_ avoid.
+                let mut params = Vec::with_capacity(chunk.len() * 4);
+                for d in chunk.iter() {
+                    params.push(
+                        d.merged_node
+                            .merge_state
+                            .local_node()
+                            .map(|node| node.guid.as_str()),
+                    );
+                    params.push(
+                        d.merged_node
+                            .merge_state
+                            .remote_node()
+                            .map(|node| node.guid.as_str()),
+                    );
+                    params.push(Some(d.merged_node.guid.as_str()));
+                    params.push(Some(d.merged_parent_node.guid.as_str()));
+                }
+                self.db.execute(&format!(
+                    "
+                    INSERT INTO mergedTree(localGuid, remoteGuid, mergedGuid, mergedParentGuid, level,
+                                           position, useRemote, shouldUpload)
+                    VALUES {}",
+                    sql_support::repeat_display(chunk.len(), ",", |index, f| {
+                        let d = &chunk[index];
+                        write!(f, "(?, ?, ?, ?, {}, {}, {}, {})",
+                            d.level, d.position, d.merged_node.merge_state.should_apply(),
+                            d.merged_node.merge_state.upload_reason() != UploadReason::None)
+                    })
+                ), &params)?;
+                Ok(())
+            },
+        )?;
+
+        // Next, insert rows for deletions. Unlike Desktop, there's no
+        // `noteItemRemoved` trigger on `itemsToRemove`, since we don't support
```

Comment:
This comment is more confusing than helpful here. Move to `create_sync_temp_tables.sql` or remove.